pkg/apis/eventing/v1alpha1/provisioner_types.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Provisioner provisions.
+type Provisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ProvisionerSpec   `json:"spec"`
+	Status ProvisionerStatus `json:"status"`
+}
+
+// Check that Provisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*Provisioner)(nil)
+var _ apis.Defaultable = (*Provisioner)(nil)
+var _ runtime.Object = (*Provisioner)(nil)
+var _ webhook.GenericCRD = (*Provisioner)(nil)
+
+// ProvisionerSpec is the spec for a Provisioner resource.
+type ProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Type is the type of the resource to be provisioned.
+	// +required
+	Type runtime.TypeMeta `json:"type"`
+}
+
+type ProvisionerConditionType string
+
+const (
+	// ProvisionerConditionReady specifies that the resource is ready.
+	ProvisionerConditionReady ProvisionerConditionType = "Ready"
+)
+
+// ProvisionerConditionStatus describes the state of this resource at a point in time.
+type ProvisionerConditionStatus struct {
+	// Type of condition.
+	// +required
+	Type ProvisionerConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status corev1.ConditionStatus `json:"status"`
+
+	// The last time this condition was updated.
+	// +optional
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// ProvisionerStatus is the status for a Provisioner resource
+type ProvisionerStatus struct {
+	// Conditions holds the state of a provisioner at a point in time.
+	Conditions []ProvisionerConditionStatus `json:"conditions,omitempty"`
+
+	// Provisioned holds the status of creation or adoption of each EventType
+	// and errors therein. It is expected that a provisioner list all produced
+	// EventTypes, if applicable.
+	Provisioned []ProvisionedStatus `json:"provisioned,omitempty"`
+
+	// ProvisionerContext is what the Provisioner operation returns and holds
+	// enough information to perform cleanup once a Provisioner is deleted.
+	// NOTE: experimental field.
+	ProvisionerContext *runtime.RawExtension `json:"provisionerContext,omitempty"`
```

Comment:
Do we need this? Where is this from? I don't understand the comment in this context.